core/storage/trie_pruner/impl/trie_pruner_impl.cpp

```diff
@@ -0,0 +1,509 @@
+    OUTCOME_TRY(child_root_hash, addNewStateWith(new_trie, version));
+    OUTCOME_TRY(markAsChild(Parent{parent_root}, key, Child{child_root_hash}));
+    return outcome::success();
+  }
+
+  outcome::result<void> TriePrunerImpl::markAsChild(Parent parent,
+                                                    const common::Buffer &key,
+                                                    Child child) {
+    std::scoped_lock lock{ref_count_mutex_};
+    child_states_[parent.hash].emplace_back(ChildStorageInfo{key, child.hash});
+    OUTCOME_TRY(savePersistentState());
+    return outcome::success();
+  }
+
+  outcome::result<storage::trie::RootHash> TriePrunerImpl::addNewStateWith(
+      const trie::PolkadotTrie &new_trie, trie::StateVersion version) {
+    if (new_trie.getRoot() == nullptr) {
+      SL_DEBUG(logger_, "Attempt to add a trie with a null root");
+      return outcome::success();
+    }
+
+    SL_DEBUG(logger_, "Ref count map size is {}", ref_count_.size());
+    KAGOME_PROFILE_START_L(logger_, register_state);
+
+    struct Entry {
+      std::shared_ptr<const trie::TrieNode> node;
+      common::Hash256 hash;
+    };
+    std::vector<Entry> queued_nodes;
+
+    EncoderCache encoder{*codec_, logger_};
+
+    OUTCOME_TRY(root_hash,
+                encoder.getMerkleValue(*new_trie.getRoot(), version));
+    BOOST_ASSERT(root_hash.isHash());
+    SL_DEBUG(logger_, "Add new state with hash: {}", root_hash.asBuffer());
+    queued_nodes.push_back({new_trie.getRoot(), *root_hash.asHash()});
+
+    std::scoped_lock lock{ref_count_mutex_};
+
+    ref_count_[*root_hash.asHash()] += 1;
+
+    size_t referenced_nodes_num = 0;
+    size_t referenced_values_num = 0;
+
+    while (!queued_nodes.empty()) {
+      auto [node, hash] = queued_nodes.back();
+      queued_nodes.pop_back();
+      SL_TRACE(logger_,
+               "Add - Node {}, ref count {}",
+               hash.toHex(),
+               ref_count_[hash]);
+
+      referenced_nodes_num++;
+      bool is_new_node_with_value = node != nullptr && ref_count_[hash] == 1
+                                 && node->getValue().is_some();
+      if (is_new_node_with_value) {
+        auto value_hash_opt = encoder.getValueHash(*node, version);
+        if (value_hash_opt) {
+          value_ref_count_[*value_hash_opt]++;
+          referenced_values_num++;
+        }
+      }
+
+      bool is_new_branch_node =
+          node != nullptr && node->isBranch() && ref_count_[hash] == 1;
+      if (is_new_branch_node) {
+        auto branch = static_cast<const trie::BranchNode *>(node.get());
+        for (auto opaque_child : branch->children) {
+          if (opaque_child != nullptr) {
+            OUTCOME_TRY(child, serializer_->retrieveNode(opaque_child));
+            OUTCOME_TRY(child_merkle_val,
+                        encoder.getMerkleValue(*child, version));
+            if (child_merkle_val.isHash()) {
+              ref_count_[*child_merkle_val.asHash()] += 1;
```

Comment:
I don't think it's possible this way because it would require two different same values to be stored by same keys, but it probably could be possible for two identical subtrees stored by different paths.